### PR TITLE
Add queries for aws provisioner new spot model

### PR DESCRIPTION
**Pull request by helfi92**
Reference: https://github.com/taskcluster/taskcluster-tools/issues/428

### CODE REVIEW

File: src/graphql/AwsProvisionerWorkerTypes.graphql [hunk 1]
```diff
@@ -194,6 +194,113 @@ type AwsProvisionerWorkerTypeState {
...
+type AwsProvisionerTerminationHealth implements Aws {
+  region: String!
+  az: String!
+  instanceType: String!
+
+  # A count of the instances which were shutdown cleanty.
+  # For the purposes of this API, a clean shutdown is one which was initiated by us.
+  # This includes API shutdowns or workers ending themselves.
+  # It does not mean the actual workload ran successfully, rather that we chose to terminate it.
+  clean_shutdown: Int!
```
Comment (eliperelman) [inline src/graphql/AwsProvisionerWorkerTypes.graphql]: I'm not sure why these are cased with underscores, but if they come from the API that way, then that is inconsistent. we should normalize this for the site so we deal with camel cased properties and not have to add ESLint overrides because of the underscores in the names.

Comment (helfi92) [reply src/graphql/AwsProvisionerWorkerTypes.graphql]: Yeah, they come from an API. I think GraphQL allows you to alias property fields. I'll do that.

Comment (eliperelman) [reply src/graphql/AwsProvisionerWorkerTypes.graphql]: Aliases are for the query, so you'll just have to do a translation mapping when querying from the data loader or mutating from the resolver.

File: src/loaders/workerTypes.js [hunk 2]
```diff
@@ -49,11 +50,41 @@ export default ({ queue, awsProvisioner }) => {
...
+        const data = await ec2Manager.getHealth();
+        const health = filter ? sift(filter, data) : data;
+        const createEntity = ({ region, az, instanceType, ...health }) =>
+          new Aws(region, az, instanceType, health);
+
+        return {
+          running: health.running.map(createEntity),
+          terminationHealth: health.terminationHealth.map(createEntity),
+          requestHealth: health.requestHealth.map(createEntity),
+        };
```
Comment (helfi92) [inline src/loaders/workerTypes.js]: Having this as an entity will require an extra 3 `map`s. Not sure if this is more efficient, but I can see how it's more readable.

File: src/resolvers/AwsProvisionerWorkerTypes.js [hunk 3]
```diff
@@ -1,9 +1,56 @@
...
+    },
+    insufficientCapacity({ insufficient_capacity }) {
+      return insufficient_capacity;
+    },
+    limitExceeded({ limit_exceeded }) {
+      return limit_exceeded;
+    },
+  },
+  /* eslint-enable camelcase */
+
```
Comment (eliperelman) [inline src/resolvers/AwsProvisionerWorkerTypes.js]: You can remove this extra line.

File: src/resolvers/AwsProvisionerWorkerTypes.js [hunk 4]
```diff
@@ -1,9 +1,56 @@
 export default {
+  /* eslint-disable camelcase */
+  AwsProvisionerTerminationHealth: {
+    cleanShutdown({ clean_shutdown }) {
```
Comment (eliperelman) [inline src/resolvers/AwsProvisionerWorkerTypes.js]: I really like this.

File: src/entities/Aws.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+export default class Aws {
+  constructor(region, az, instanceType, data) {
```
Comment (eliperelman) [inline src/entities/Aws.js]: It looks like these 3 properties don't get any special treatment over the rest of the data, and I only mentioned creating this entity because I thought we were going to map the underscored properties in here. Since you found a better way to do it, we probably don't need this entity.

File: src/loaders/workerTypes.js [hunk 6]
```diff
@@ -49,11 +50,41 @@ export default ({ queue, awsProvisioner }) => {
...
+  const awsProvisionerHealth = new DataLoader(queries =>
+    Promise.all(
+      queries.map(async ({ filter }) => {
+        const data = await ec2Manager.getHealth();
+        const health = filter ? sift(filter, data) : data;
+        const createEntity = ({ region, az, instanceType, ...health }) =>
+          new Aws(region, az, instanceType, health);
+
+        return {
+          running: health.running.map(createEntity),
```
Comment (eliperelman) [inline src/loaders/workerTypes.js]: Yeah, since you aren't doing the underscore mapping in the entity, this no longer has value. You should be able to pass these directly now:

```js
const health = await ec2Manager.getHealth();

return filter ? sift(filter, health) : health;
```

File: src/entities/Aws.js (hunk 5 above)
Comment (helfi92) [reply src/entities/Aws.js]: ++ Agreed.

**Comment by helfi92** [comment]
Dropped the commit that created the entity.
